udev.7
.\"     Title: udev
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.73.2 <http://docbook.sf.net/>
.\"      Date: August 2005
.\"    Manual: udev
.\"    Source: udev
.\"
.TH "UDEV" "7" "August 2005" "udev" "udev"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
udev - dynamic device management
.SH "DESCRIPTION"
.PP
udev provides a dynamic device directory containing only the files for actually present devices\. It creates or removes device node files in the
\fI/dev\fR
directory, or it renames network interfaces\.
.PP
Usually udev runs as
\fBudevd\fR(8)
and receives uevents directly from the kernel if a device is added or removed from the system\.
.PP
If udev receives a device event, it matches its configured rules against the available device attributes provided in sysfs to identify the device\. Rules that match may provide additional device information or specify a device node name and multiple symlink names and instruct udev to run additional programs as part of the device event handling\.
.SH "CONFIGURATION"
.PP
udev configuration files are placed in
\fI/etc/udev/\fR
and
\fI/lib/udev/\fR\. All empty lines, or lines beginning with \'#\' will be ignored\.
.SS "Configuration file"
.PP
udev expects its main configuration file at
\fI/etc/udev/udev\.conf\fR\. It consists of a set of variables allowing the user to override default udev values\. The following variables can be set:
.PP
\fBudev_root\fR
.RS 4
Specifies where to place the device nodes in the filesystem\. The default value is
\fI/dev\fR\.
.RE
.PP
\fBudev_log\fR
.RS 4
The logging priority\. Valid values are the numerical syslog priorities or their textual representations:
\fBerr\fR,
\fBinfo\fR
and
\fBdebug\fR\.
.RE
.SS "Rules files"
.PP
The udev rules are read from the files located in the default rules directory
\fI/lib/udev/rules\.d/\fR, the custom rules directory
\fI/etc/udev/rules\.d/\fR
and the temporary rules directory
\fI/dev/\.udev/rules\.d/\fR\. All rule files are sorted and processed in lexical order, regardless in which of these directories they live\. Every line in the rules file contains at least one key value pair\. There are two kind of keys, match and assignment keys\. If all match keys are matching against its value, the rule gets applied and the assign keys get the specified value assigned\.
.PP
A matching rule may specify the name of the device node, add a symlink pointing to the node, or run a specified program as part of the event handling\. If no matching rule is found, the default device node name is used\.
.PP
A rule may consist of a list of one or more key value pairs separated by a comma\. Each key has a distinct operation, depending on the used operator\. Valid operators are:
.PP
\fB==\fR
.RS 4
Compare for equality\.
.RE
.PP
\fB!=\fR
.RS 4
Compare for non\-equality\.
.RE
.PP
\fB=\fR
.RS 4
Assign a value to a key\. Keys that represent a list, are reset and only this single value is assigned\.
.RE
.PP
\fB+=\fR
.RS 4
Add the value to a key that holds a list of entries\.
.RE
.PP
\fB:=\fR
.RS 4
Assign a value to a key finally; disallow any later changes, which may be used to prevent changes by any later rules\.
.RE
.PP
The following key names can be used to match against device properties\. Some of the keys also match against properties of the parent devices in sysfs, not only the device that has generated the event\. If multiple keys that match a parent device are specified in a single rule, all these keys must match at one and the same parent device\.
.PP
\fBACTION\fR
.RS 4
Match the name of the event action\.
.RE
.PP
\fBDEVPATH\fR
.RS 4
Match the devpath of the event device\.
.RE
.PP
\fBKERNEL\fR
.RS 4
Match the name of the event device\.
.RE
.PP
\fBNAME\fR
.RS 4
Match the name of the node or network interface\. It can be used once the NAME key has been set in one of the preceding rules\.
.RE
.PP
\fBSUBSYSTEM\fR
.RS 4
Match the subsystem of the event device\.
.RE
.PP
\fBDRIVER\fR
.RS 4
Match the driver name of the event device\. Only set for devices which are bound to a driver at the time the event is generated\.
.RE
.PP
\fBATTR{\fR\fB\fIfilename\fR\fR\fB}\fR
.RS 4
Match sysfs attribute values of the event device\. Up to five
\fBATTR\fR
keys can be specified per rule\. Trailing whitespace in the attribute values is ignored, if the specified match value does not contain trailing whitespace itself\. Depending on the type of operator, this key is also used to set the value of a sysfs attribute\.
.RE
.PP
\fBKERNELS\fR
.RS 4
Search the devpath upwards for a matching device name\.
.RE
.PP
\fBSUBSYSTEMS\fR
.RS 4
Search the devpath upwards for a matching device subsystem name\.
.RE
.PP
\fBDRIVERS\fR
.RS 4
Search the devpath upwards for a matching device driver name\.
.RE
.PP
\fBATTRS{\fR\fB\fIfilename\fR\fR\fB}\fR
.RS 4
Search the devpath upwards for a device with matching sysfs attribute values\. Up to five
\fBATTRS\fR
keys can be specified per rule, but all of them must match on the same device\. Trailing whitespace in the attribute values is ignored, if the specified match value does not contain trailing whitespace itself\.
.RE
.PP
\fBENV{\fR\fB\fIkey\fR\fR\fB}\fR
.RS 4
Match against the value of an environment variable\. Up to five
\fBENV\fR
keys can be specified per rule\. Depending on the type of operator, this key is also used to export a variable to the environment\.
.RE
.PP
\fBTEST{\fR\fB\fIoctal mode mask\fR\fR\fB}\fR
.RS 4
Test the existence of a file\. An octal mode mask can be specified if needed\.
.RE
.PP
\fBPROGRAM\fR
.RS 4
Execute external program\. The key is true, if the program returns with exit code zero\. The whole event environment is available to the executed program\. The program\'s output printed to stdout, is available in the RESULT key\.
.RE
.PP
\fBRESULT\fR
.RS 4
Match the returned string of the last PROGRAM call\. This key can be used in the same or in any later rule after a PROGRAM call\.
.RE
.PP
Most of the fields support a shell style pattern matching\. The following pattern characters are supported:
.PP
\fB*\fR
.RS 4
Matches zero, or any number of characters\.
.RE
.PP
\fB?\fR
.RS 4
Matches any single character\.
.RE
.PP
\fB[]\fR
.RS 4
Matches any single character specified within the brackets\. For example, the pattern string \'tty[SR]\' would match either \'ttyS\' or \'ttyR\'\. Ranges are also supported within this match with the \'\-\' character\. For example, to match on the range of all digits, the pattern [0\-9] would be used\. If the first character following the \'[\' is a \'!\', any characters not enclosed are matched\.
.RE
.PP
The following keys can get values assigned:
.PP
\fBNAME\fR
.RS 4
The name of the node to be created, or the name the network interface should be renamed to\. Only one rule can set the node name, all later rules with a NAME key will be ignored\.
.RE
.PP
\fBSYMLINK\fR
.RS 4
The name of a symlink targeting the node\. Every matching rule can add this value to the list of symlinks to be created along with the device node\. Multiple symlinks may be specified by separating the names by the space character\.
.RE
.PP
\fBOWNER, GROUP, MODE\fR
.RS 4
The permissions for the device node\. Every specified value overwrites the compiled\-in default value\.
.RE
.PP
\fBATTR{\fR\fB\fIkey\fR\fR\fB}\fR
.RS 4
The value that should be written to a sysfs attribute of the event device\. Depending on the type of operator, this key is also used to match against the value of a sysfs attribute\.
.RE
.PP
\fBENV{\fR\fB\fIkey\fR\fR\fB}\fR
.RS 4
Export a variable to the environment\. Depending on the type of operator, this key is also to match against an environment variable\.
.RE
.PP
\fBRUN\fR
.RS 4
Add a program to the list of programs to be executed for a specific device\. This can only be used for very short running tasks\. Running an event process for a long period of time may block all further events for this or a dependent device\. Long running tasks need to be immediately detached from the event process itself\.
.sp
If the specifiefd string starts with
\fBsocket:\fR\fB\fIpath\fR\fR, all current event values will be passed to the specified socket, as a message in the same format the kernel sends an uevent\. If the first character of the specified path is an @ character, an abstract namespace socket is used, instead of an existing socket file\.
.RE
.PP
\fBLABEL\fR
.RS 4
Named label where a GOTO can jump to\.
.RE
.PP
\fBGOTO\fR
.RS 4
Jumps to the next LABEL with a matching name
.RE
.PP
\fBIMPORT{\fR\fB\fItype\fR\fR\fB}\fR
.RS 4
Import a set of variables into the event environment, depending on
\fItype\fR:
.PP
\fBprogram\fR
.RS 4
Execute an external program specified as the assigned value and import its output, which must be in environment key format\.
.RE
.PP
\fBfile\fR
.RS 4
Import a text file specified as the assigned value, which must be in environment key format\.
.RE
.PP
\fBparent\fR
.RS 4
Import the stored keys from the parent device by reading the database entry of the parent device\. The value assigned to
\fBIMPORT{parent}\fR
is used as a filter of key names to import (with the same shell\-style pattern matching used for comparisons)\.
.RE
.sp
If no option is given, udev will choose between
\fBprogram\fR
and
\fBfile\fR
based on the executable bit of the file permissions\.
.RE
.PP
\fBWAIT_FOR\fR
.RS 4
Wait for a file to become available\.
.RE
.PP
\fBOPTIONS\fR
.RS 4
Rule and device options:
.PP
\fBlast_rule\fR
.RS 4
Stops further rules application\. No later rules will have any effect\.
.RE
.PP
\fBignore_device\fR
.RS 4
Ignore this event completely\.
.RE
.PP
\fBignore_remove\fR
.RS 4
Do not remove the device node when the device goes away\. This may be useful as a workaround for broken device drivers\.
.RE
.PP
\fBlink_priority=\fR\fB\fIvalue\fR\fR
.RS 4
Specify the priority of the created symlinks\. Devices with higher priorities overwrite existing symlinks of other devices\. The default is 0\.
.RE
.PP
\fBall_partitions\fR
.RS 4
Create the device nodes for all available partitions of a block device\. This may be useful for removable media devices where media changes are not detected\.
.RE
.PP
\fBevent_timeout=\fR
.RS 4
Number of seconds an event will wait for operations to finish, before it will terminate itself\.
.RE
.PP
\fBstring_escape=\fR\fB\fInone|replace\fR\fR
.RS 4
Usually control and other possibly unsafe characters are replaced in strings used for device naming\. The mode of replacement can be specified with this option\.
.RE
.RE
.PP
The
\fBNAME\fR,
\fBSYMLINK\fR,
\fBPROGRAM\fR,
\fBOWNER\fR,
\fBGROUP\fR,
\fBMODE\fR
and
\fBRUN\fR
fields support simple printf\-like string substitutions\. The
\fBRUN\fR
format chars gets applied after all rules have been processed, right before the program is executed\. It allows the use of the complete environment set by earlier matching rules\. For all other fields, substitutions are applied while the individual rule is being processed\. The available substitutions are:
.PP
\fB$kernel\fR, \fB%k\fR
.RS 4
The kernel name for this device\.
.RE
.PP
\fB$number\fR, \fB%n\fR
.RS 4
The kernel number for this device\. For example, \'sda3\' has kernel number of \'3\'
.RE
.PP
\fB$devpath\fR, \fB%p\fR
.RS 4
The devpath of the device\.
.RE
.PP
\fB$id\fR, \fB%b\fR
.RS 4
The name of the device matched while searching the devpath upwards for
\fBSUBSYSTEMS\fR,
\fBKERNELS\fR,
\fBDRIVERS\fR
and
\fBATTRS\fR\.
.RE
.PP
\fB$driver\fR
.RS 4
The driver name of the device matched while searching the devpath upwards for
\fBSUBSYSTEMS\fR,
\fBKERNELS\fR,
\fBDRIVERS\fR
and
\fBATTRS\fR\.
.RE
.PP
\fB$attr{\fR\fB\fIfile\fR\fR\fB}\fR, \fB%s{\fR\fB\fIfile\fR\fR\fB}\fR
.RS 4
The value of a sysfs attribute found at the device, where all keys of the rule have matched\. If the matching device does not have such an attribute, follow the chain of parent devices and use the value of the first attribute that matches\. If the attribute is a symlink, the last element of the symlink target is returned as the value\.
.RE
.PP
\fB$env{\fR\fB\fIkey\fR\fR\fB}\fR, \fB%E{\fR\fB\fIkey\fR\fR\fB}\fR
.RS 4
The value of an environment variable\.
.RE
.PP
\fB$major\fR, \fB%M\fR
.RS 4
The kernel major number for the device\.
.RE
.PP
\fB$minor\fR, \fB%m\fR
.RS 4
The kernel minor number for the device\.
.RE
.PP
\fB$result\fR, \fB%c\fR
.RS 4
The string returned by the external program requested with PROGRAM\. A single part of the string, separated by a space character may be selected by specifying the part number as an attribute:
\fB%c{N}\fR\. If the number is followed by the \'+\' char this part plus all remaining parts of the result string are substituted:
\fB%c{N+}\fR
.RE
.PP
\fB$parent\fR, \fB%P\fR
.RS 4
The node name of the parent device\.
.RE
.PP
\fB$name\fR
.RS 4
The current name of the device node\. If not changed by a rule, it is the name of the kernel device\.
.RE
.PP
\fB$links\fR
.RS 4
The current list of symlinks, separated by a space character\. The value is only set if an earlier rule assigned a value, or during a remove events\.
.RE
.PP
\fB$root\fR, \fB%r\fR
.RS 4
The udev_root value\.
.RE
.PP
\fB$sys\fR, \fB%S\fR
.RS 4
The sysfs mount point\.
.RE
.PP
\fB$tempnode\fR, \fB%N\fR
.RS 4
The name of a created temporary device node to provide access to the device from a external program before the real node is created\.
.RE
.PP
\fB%%\fR
.RS 4
The \'%\' character itself\.
.RE
.PP
\fB$$\fR
.RS 4
The \'$\' character itself\.
.RE
.PP
The count of characters to be substituted may be limited by specifying the format length value\. For example, \'%3s{file}\' will only insert the first three characters of the sysfs attribute
.SH "AUTHOR"
.PP
Written by Greg Kroah\-Hartman
<greg@kroah\.com>
and Kay Sievers
<kay\.sievers@vrfy\.org>\. With much help from Dan Stekloff and many others\.
.SH "SEE ALSO"
.PP
\fBudevd\fR(8),
\fBudevadm\fR(8)


udev.c
/*
 * udev.c
 *
 * Copyright (C) 2003-2004 Greg Kroah-Hartman <greg@kroah.com>
 * Copyright (C) 2004-2005 Kay Sievers <kay.sievers@vrfy.org>
 *
 *	This program is free software; you can redistribute it and/or modify it
 *	under the terms of the GNU General Public License as published by the
 *	Free Software Foundation version 2 of the License.
 * 
 *	This program is distributed in the hope that it will be useful, but
 *	WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *	General Public License for more details.
 * 
 *	You should have received a copy of the GNU General Public License along
 *	with this program; if not, write to the Free Software Foundation, Inc.,
 *	675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <ctype.h>
#include <errno.h>
#include <signal.h>
#include <unistd.h>
#include <syslog.h>

#include "udev.h"
#include "udev_rules.h"
#include "udev_selinux.h"

#ifdef USE_LOG
void log_message(int priority, const char *format, ...)
{
	va_list args;

	if (priority > udev_log_priority)
		return;

	va_start(args, format);
	vsyslog(priority, format, args);
	va_end(args);
}
#endif

static void asmlinkage sig_handler(int signum)
{
	switch (signum) {
		case SIGALRM:
			exit(1);
		case SIGINT:
		case SIGTERM:
			exit(20 + signum);
	}
}

int main(int argc, char *argv[], char *envp[])
{
	struct sysfs_device *dev;
	struct udevice *udev;
	const char *maj, *min;
	struct udev_rules rules;
	const char *action;
	const char *devpath;
	const char *subsystem;
	struct sigaction act;
	int devnull;
	int retval = -EINVAL;

	if (argc == 2 && strcmp(argv[1], "-V") == 0) {
		printf("%s\n", UDEV_VERSION);
		exit(0);
	}

	/* set std fd's to /dev/null, /sbin/hotplug forks us, we don't have them at all */
	devnull = open("/dev/null", O_RDWR);
	if (devnull >= 0)  {
		if (devnull != STDIN_FILENO)
			dup2(devnull, STDIN_FILENO);
		if (devnull != STDOUT_FILENO)
			dup2(devnull, STDOUT_FILENO);
		if (devnull != STDERR_FILENO)
			dup2(devnull, STDERR_FILENO);
		if (devnull > STDERR_FILENO)
			close(devnull);
	}

	logging_init("udev");
	if (devnull < 0)
		err("open /dev/null failed: %s", strerror(errno));
	udev_config_init();
	selinux_init();
	dbg("version %s", UDEV_VERSION);

	/* set signal handlers */
	memset(&act, 0x00, sizeof(act));
	act.sa_handler = (void (*)(int)) sig_handler;
	sigemptyset (&act.sa_mask);
	act.sa_flags = 0;
	sigaction(SIGALRM, &act, NULL);
	sigaction(SIGINT, &act, NULL);
	sigaction(SIGTERM, &act, NULL);

	/* trigger timeout to prevent hanging processes */
	alarm(UDEV_ALARM_TIMEOUT);

	action = getenv("ACTION");
	devpath = getenv("DEVPATH");
	subsystem = getenv("SUBSYSTEM");
	/* older kernels passed the SUBSYSTEM only as argument */
	if (subsystem == NULL && argc == 2)
		subsystem = argv[1];

	if (action == NULL || subsystem == NULL || devpath == NULL) {
		err("action, subsystem or devpath missing");
		goto exit;
	}

	/* export log_priority , as called programs may want to do the same as udev */
	if (udev_log_priority) {
		char priority[32];

		sprintf(priority, "%i", udev_log_priority);
		setenv("UDEV_LOG", priority, 1);
	}

	sysfs_init();
	udev_rules_init(&rules, 0);

	dev = sysfs_device_get(devpath);
	if (dev == NULL) {
		info("unable to open '%s'", devpath);
		goto fail;
	}

	udev = udev_device_init();
	if (udev == NULL)
		goto fail;

	/* override built-in sysfs device */
	udev->dev = dev;
	strlcpy(udev->action, action, sizeof(udev->action));

	/* get dev_t from environment, which is needed for "remove" to work, "add" works also from sysfs */
	maj = getenv("MAJOR");
	min = getenv("MINOR");
	if (maj != NULL && min != NULL)
		udev->devt = makedev(atoi(maj), atoi(min));
	else
		udev->devt = udev_device_get_devt(udev);

	retval = udev_device_event(&rules, udev);

	if (retval == 0 && !udev->ignore_device && udev_run) {
		struct name_entry *name_loop;

		dbg("executing run list");
		list_for_each_entry(name_loop, &udev->run_list, node) {
			if (strncmp(name_loop->name, "socket:", strlen("socket:")) == 0)
				pass_env_to_socket(&name_loop->name[strlen("socket:")], devpath, action);
			else {
				char program[PATH_SIZE];

				strlcpy(program, name_loop->name, sizeof(program));
				udev_rules_apply_format(udev, program, sizeof(program));
				run_program(program, udev->dev->subsystem, NULL, 0, NULL, (udev_log_priority >= LOG_INFO));
			}
		}
	}

	udev_device_cleanup(udev);
fail:
	udev_rules_cleanup(&rules);
	sysfs_cleanup();

exit:
	logging_close();
	if (retval != 0)
		return 1;
	return 0;
}
